pkg/tso/allocator.go

```diff
@@ -0,0 +1,433 @@
+			return
+		default:
+		}
+
+		primary, checkAgain := a.member.CheckLeader()
+		if checkAgain {
+			continue
+		}
+		if primary != nil {
+			log.Info("start to watch the primary",
+				append(a.logFields, zap.Stringer("tso-primary", primary))...)
+			// Watch will keep looping and never return unless the primary has changed.
+			primary.Watch(a.ctx)
+			log.Info("the tso primary has changed, try to re-campaign a primary",
+				append(a.logFields, zap.Stringer("old-tso-primary", primary))...)
+		}
+
+		// To make sure the expected primary(if existed) and new primary are on the same server.
+		expectedPrimary := mcsutils.GetExpectedPrimaryFlag(a.member.Client(), &keypath.MsParam{
+			ServiceName: constant.TSOServiceName,
+			GroupID:     a.keyspaceGroupID,
+		})
+		// skip campaign the primary if the expected primary is not empty and not equal to the current memberValue.
+		// expected primary ONLY SET BY `{service}/primary/transfer` API.
+		if len(expectedPrimary) > 0 && !strings.Contains(a.member.MemberValue(), expectedPrimary) {
+			log.Info("skip campaigning of tso primary and check later", append(a.logFields,
+				zap.String("expected-primary-id", expectedPrimary),
+				zap.String("cur-member-value", a.member.MemberValue()))...)
+			time.Sleep(200 * time.Millisecond)
+			continue
+		}
+
+		a.campaignLeader()
+	}
+}
+
+func (a *Allocator) campaignLeader() {
+	log.Info("start to campaign the primary", a.logFields...)
+	leaderLease := a.cfg.GetLeaderLease()
+	if err := a.member.CampaignLeader(a.ctx, leaderLease); err != nil {
+		if errors.Is(err, errs.ErrEtcdTxnConflict) {
+			log.Info("campaign tso primary meets error due to txn conflict, another tso server may campaign successfully",
+				a.logFields...)
+		} else if errors.Is(err, errs.ErrCheckCampaign) {
+			log.Info("campaign tso primary meets error due to pre-check campaign failed, the tso keyspace group may be in split",
+				a.logFields...)
+		} else {
+			log.Error("campaign tso primary meets error due to etcd error", append(a.logFields, errs.ZapError(err))...)
+		}
+		return
+	}
+
+	// Start keepalive the leadership and enable TSO service.
+	// TSO service is strictly enabled/disabled by the leader lease for 2 reasons:
+	//   1. lease based approach is not affected by thread pause, slow runtime schedule, etc.
+	//   2. load region could be slow. Based on lease we can recover TSO service faster.
+	ctx, cancel := context.WithCancel(a.ctx)
+	var resetLeaderOnce sync.Once
+	defer resetLeaderOnce.Do(func() {
+		cancel()
+		a.member.ResetLeader()
+	})
+
+	// maintain the leadership, after this, TSO can be service.
+	a.member.KeepLeader(ctx)
+	log.Info("campaign tso primary ok", a.logFields...)
+
+	log.Info("initializing the tso allocator")
+	if err := a.Initialize(); err != nil {
+		log.Error("failed to initialize the tso allocator", append(a.logFields, errs.ZapError(err))...)
+		return
+	}
+	defer func() {
+		// Leader will be reset in `resetLeaderOnce` later.
+		a.Reset(false)
```

Comment:
Previously, we reset the leader here.

Reply:
As mentioned in the comment, the leader will be reset in `resetLeaderOnce` later in the previous defer.